File: solutions/topological-sort/329-longest.py
I'll help you implement a solution for the Longest Increasing Path problem in Python. This solution will use DFS with memoization to find the longest increasing path in a matrix.

```python
from typing import List

class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        """
        Find the longest increasing path in a matrix where you can move in four directions
        (up, down, left, right).
        
        Args:
            matrix: 2D list of integers representing the matrix
            
        Returns:
            int: Length of the longest increasing path
        """
        if not matrix or not matrix[0]:
            return 0
        
        rows, cols = len(matrix), len(matrix[0])
        # Cache to store computed results for each cell
        memo = {}
        
        def dfs(row: int, col: int, prev_val: int) -> int:
            """
            Depth-first search helper function to explore paths.
            
            Args:
                row: Current row position
                col: Current column position
                prev_val: Value of the previous cell
                
            Returns:
                int: Length of the longest path starting from current position
            """
            # Check boundaries and if current value is less than previous
            if (row < 0 or row >= rows or 
                col < 0 or col >= cols or 
                matrix[row][col] <= prev_val):
                return 0
            
            # If result is cached, return it
            if (row, col) in memo:
                return memo[(row, col)]
            
            current = matrix[row][col]
            
            # Explore all four directions
            up = dfs(row - 1, col, current)
            down = dfs(row + 1, col, current)
            left = dfs(row, col - 1, current)
            right = dfs(row, col + 1, current)
            
            # Store the maximum path length starting from current cell
            memo[(row, col)] = 1 + max(up, down, left, right)
            return memo[(row, col)]
        
        # Try starting from each cell and find the maximum path length
        max_path = 0
        for i in range(rows):
            for j in range(cols):
                max_path = max(max_path, dfs(i, j, float('-inf')))
        
        return max_path


def test_longest_increasing_path():
    """
    Test function to verify the implementation
    """
    # Test cases
    solution = Solution()
    
    # Test case 1
    matrix1 = [
        [9,9,4],
        [6,6,8],
        [2,1,1]
    ]
    assert solution.longestIncreasingPath(matrix1) == 4
    
    # Test case 2
    matrix2 = [
        [3,4,5],
        [3,2,6],
        [2,2,1]
    ]
    assert solution.longestIncreasingPath(matrix2) == 4
    
    # Test case 3: Empty matrix
    matrix3 = []
    assert solution.longestIncreasingPath(matrix3) == 0
    
    # Test case 4: Single element matrix
    matrix4 = [[1]]
    assert solution.longestIncreasingPath(matrix4) == 1
    
    print("All test cases passed!")


if __name__ == "__main__":
    test_longest_increasing_path()
```

This implementation includes:

1. A `Solution` class with the main `longestIncreasingPath` method
2. A DFS helper function with memoization to avoid redundant calculations
3. Comprehensive handling of edge cases (empty matrix, single element matrix)
4. Test cases to verify the implementation
5. Clear comments explaining the logic and functionality
6. Type hints for better code readability and maintainability

The algorithm works by:
1. Using DFS to explore all possible paths from each cell
2. Memoizing results to avoid recalculating paths
3. Checking all four directions (up, down, left, right) for each cell
4. Maintaining a cache of computed results for efficiency
5. Returning the maximum path length found

The time complexity is O(m*n) where m and n are the dimensions of the matrix, and space complexity is O(m*n) for the memoization cache.

The code follows Python best practices and includes proper error handling and edge cases. It can be run directly to test the implementation with the included test cases.

File: solutions/topological-sort/329-longest.js
I'll help you create a solution for the Longest Increasing Path problem in JavaScript. I'll write a complete implementation with comments and proper error handling.

```javascript
/**
 * Longest Increasing Path in a Matrix
 * Time Complexity: O(m * n) where m and n are the dimensions of the matrix
 * Space Complexity: O(m * n) for the memoization cache
 * 
 * @param {number[][]} matrix - The input matrix of integers
 * @return {number} - Length of the longest increasing path
 */
function longestIncreasingPath(matrix) {
    // Handle edge cases
    if (!matrix || matrix.length === 0 || matrix[0].length === 0) {
        return 0;
    }

    const rows = matrix.length;
    const cols = matrix[0].length;
    
    // Create memoization cache
    const memo = Array(rows).fill().map(() => Array(cols).fill(0));
    
    // Track maximum path length
    let maxPath = 0;
    
    // Direction arrays for exploring adjacent cells
    const directions = [
        [-1, 0],  // up
        [1, 0],   // down
        [0, -1],  // left
        [0, 1]    // right
    ];
    
    /**
     * DFS helper function to explore paths starting from a cell
     * @param {number} row - Current row
     * @param {number} col - Current column
     * @param {number} prevVal - Value of the previous cell
     * @returns {number} - Length of the longest path from current cell
     */
    function dfs(row, col, prevVal) {
        // Check boundaries and if current value is less than or equal to previous
        if (row < 0 || row >= rows || col < 0 || col >= cols || matrix[row][col] <= prevVal) {
            return 0;
        }
        
        // If we have already computed this path, return cached result
        if (memo[row][col] !== 0) {
            return memo[row][col];
        }
        
        let currentMax = 1;
        
        // Explore all four directions
        for (const [dx, dy] of directions) {
            const newRow = row + dx;
            const newCol = col + dy;
            
            const pathLength = 1 + dfs(newRow, newCol, matrix[row][col]);
            currentMax = Math.max(currentMax, pathLength);
        }
        
        // Cache the result
        memo[row][col] = currentMax;
        return currentMax;
    }
    
    // Try starting from each cell in the matrix
    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            maxPath = Math.max(maxPath, dfs(i, j, -Infinity));
        }
    }
    
    return maxPath;
}

/**
 * Test cases
 */
function runTests() {
    const testCases = [
        {
            input: [
                [9,9,4],
                [6,6,8],
                [2,1,1]
            ],
            expected: 4
        },
        {
            input: [
                [3,4,5],
                [3,2,6],
                [2,2,1]
            ],
            expected: 4
        },
        {
            input: [[1]],
            expected: 1
        },
        {
            input: [],
            expected: 0
        }
    ];

    for (let i = 0; i < testCases.length; i++) {
        const result = longestIncreasingPath(testCases[i].input);
        console.log(`Test case ${i + 1}:`);
        console.log(`Input: ${JSON.stringify(testCases[i].input)}`);
        console.log(`Expected: ${testCases[i].expected}`);
        console.log(`Got: ${result}`);
        console.log(`Pass: ${result === testCases[i].expected}\n`);
    }
}

// Run tests if not being imported as a module
if (require.main === module) {
    runTests();
}

// Export for use in other modules
module.exports = longestIncreasingPath;
```

This implementation provides a solution to find the longest increasing path in a matrix. Here are the key features:

1. The main function `longestIncreasingPath` takes a matrix as input and returns the length of the longest increasing path.

2. It uses a depth-first search (DFS) approach with memoization to optimize performance.

3. Key optimizations:
   - Memoization cache to avoid recalculating paths
   - Direction array for clean navigation of adjacent cells
   - Early boundary and value checks

4. The solution includes:
   - Comprehensive error handling
   - Clear comments explaining the implementation
   - Test cases with various scenarios
   - Proper module exports

5. Time complexity is O(m * n) where m and n are the dimensions of the matrix
   Space complexity is O(m * n) for the memoization cache

6. The code includes a test suite that can be run directly or ignored when imported as a module.

To use this code, you can either:
1. Run it directly to execute the test cases
2. Import it as a module in another file
3. Call the `longestIncreasingPath` function with your own matrix input

The implementation follows JavaScript best practices and conventions while maintaining readability and efficiency.